CanvasColorSpaceProposal.md

```diff
@@ -63,56 +104,72 @@ partial interface CanvasRenderingContext2D {
 
 Example:
 <pre>
-canvas.getContext('2d', { colorSpace: "linear-srgb", pixelFormat: "float16"});
+canvas.getContext('2d', { colorSpace: "rec2020",
+                          colorEncoding: "float16",
+                          colorLinearity: "radiance"} );
 </pre>
 
-#### Common properties of all color spaces
+#### The colorSpace canvas creation parameter
+
+* Color spaces match their respective counterparts as defined in the [CSS colorspaces](https://www.w3.org/TR/css-color-4/#predefined).
 * All input colors (e.g, fillStyle or strokeStyle, and gradient stops) follow the same interpretation as CSS color literals, regardless of canvas color space.
-* Images with no color profile, when drawn to the canvas, are assumed to already be in sRGB color space.
-* Unless otherwise explicitly specified by the user, toDataURL/toBlob will produce resources in sRGB color space. If the encoding format supports colorspace tagging or embedded color profiles, the resource will be tagged as being in sRGB color space.
-* Values written to WebGL drawing buffers (e.g, values written to gl_FragColor or the clear color) are defined to be in the output canvas color space.
+* Values written to WebGL backbuffers (e.g, values written to gl_FragColor or the clear color) are in the canvas's color space and linearity.
+* Images with no color profile, when drawn to the canvas, are assumed to be in the sRGB color space.
+* Unless otherwise explicitly specified by the user, toDataURL/toBlob will produce resources in sRGB color space, with unorm8 encoding (matching existing behavior), and "luminance" linearity. If the destination image format supports colorspace tagging or embedded color profiles, the resource will be tagged as being in sRGB color space.
```

Comment:
I had the same question. Should this combination be disallowed, and if so, how? https://html.spec.whatwg.org/multipage/canvas.html#dom-canvas-getcontext is specified to return the same context if one was fetched previously, no matter whether the context creation attributes are legal, so detecting whether a combination is legal should probably not be delegated to getContext(), but rather to some other API.

Reply:
Color space and encoding are technically orthogonal, so that would work fine. That'd be using rec2020 chromaticities (primaries) and (output) transfer function (gamma), but encoding 0.5 as 0xbc. 

Color space includes both chromaticities (primaries and white point) and transfer function, not just RGB primaries, so color space is more accurate IMO. Rec2020 (and friends) do also define a preferred encoding, but we can ignore it for canvas. (it would be prohibitively slow for canvas)

Reply:
> Color space includes both chromaticities (primaries and white point) and transfer function

Then what on earth is the "linearity" attribute all about?

Reply:
> That'd be using rec2020 chromaticities (primaries) and (output) transfer function (gamma), but encoding 0.5 as 0xbc.

Sorry, I'm not folowing. Rec.2020 defines its own non-linear transfer function ("gamma" curve) that is identical to Rec.709 but slightly different from sRGB. Are you saying that such configuration would imply applying non-linear transfer twice? What's the use-case?

For documentation purposes, we could split fields even further:
- color primaries, i.e. RGB-CIE relations:
  - sRGB/Rec.709
  - DCI P3
  - Rec.2020
- transfer function (precision redistribution), three distinct options:
  - linear
  - non-linear sRGB
  - non-linear Rec.709 / Rec.2020
- data type
  - unorm8
  - unorm16
  - float16

Of course, not all combinations make practical sense so only few of them should be allowed in the spec. In my opinion, these options should probably be enough:

**Legacy**
- sRGB primaries / linear transfer / unorm8

**Correct 8-bit backbuffer**
- sRGB primaries / sRGB transfer / unorm8 

**Display P3**
- P3 primaries / sRGB transfer / unorm16

**Rec.2020**
- Rec.2020 primaries / Rec.2020 transfer / ??? (10 or 12 bit storage)

**Unbiased HDR**
- Rec.2020 primaries / linear transfer / float16

Reply:
This is a lot of options to provide initially. Can you think of a way to whittle down this list to something more feasible to implement initially, while leaving open the possibility to extend the capability set later? From one browser implementer's standpoint, I have no idea what supporting the P3 and Rec.2020 primaries implies from the browser's compositor and from the hardware's overlay plane support.

Perhaps focus on just enabling the "unorm8-srgb" and "float16" encodings with whatever other knobs are absolutely necessary.

Reply:
> Legacy
>
>    sRGB primaries / linear transfer / unorm8

I know nothing about canvas 2D implementations but the WebGL canvas is sRGB both in primaries and transfer function (canvas bits to the eyes). The only thing that thinks differently is WebGL which labels it linear thus turning off the GPU's sRGB helpers.

Reply:
>  The only thing that thinks differently is WebGL ...

That's what I called "Legacy". From an application POV, the backbuffer doesn't have a transfer function.

>  Can you think of a way to whittle down this list to something more feasible to implement initially,

That's exactly the point of providing fixed presets in well-defined terms instead of independent knobs to a user. This bullets of this list could be implemented step-by-step, as the ecosystem advances.

- "Legacy" - available today
- "Correct 8-bit backbuffer" - highly-anticipated sRGB-correct backbuffer
- "Display P3" - configuration natively supported by the latest high-end mobile devices
- "Rec.2020" - no easy path today, just an example for future
- "Unbiased HDR" - most-flexible, implementation-dependent